### PR TITLE
108 map interaction rework

### LINKED ISSUE #108: Rework interacting with map
Everything on the map is a Tile (interface). TileType should let have a parameter of which Tile implementation. Group determines has default values for each group

# Tile 
* Where it is
* What phases it is active
* What priority it has (sorted by after phase)
* What it should do (returns the type of the tile)
* What TileType it is

## Attributes
Attributes describes how a tile should behave and when. 

* Active phases
* Can be coloured
* Can be healed
* Can move
* Direction
* Collidable (direction determines where)
* Active During Phase (if a tile should do its action every time something is moved onto it)
* Active After Phase (enabling both will make it always active)

## Holes
* If a tile can move it will be killed
* Always active 

## Wall
* Has a direction it blocks (ie NORTH blocks moveables from moving to the from from it)
* Always active

## Conveyor
* Moves moveable that is placed on it
* Decrease movement point of moveable moved and call `action` on the next tile
* Active after phase

## Moveable
* Each moveable know how far conveyors should move them (get told thing when phase ended)

## Player 
* Color
* Moveable 
* Active during phase

### PR DESCRIPTION
**Pull request by kh498**
Fixes #108

### CODE REVIEW

**Comment by DanielBerge** [comment]
Fixes #91

File: src/test/java/no/uib/inf112/core/map/ConveryorInteractOnUserTest.java [hunk 1]
```diff
@@ -1,42 +1,34 @@
...
+//        testBot = roboRally.getPlayerHandler().testPlayer();
+//    }
+//
+//    private void moveTestBotTo(int x, int y) {
+//        roboX = x;
+//        roboY = y;
+//        testBot.teleport(x, y);
+//    }
+//
+//    //TODO issue 100, all these tests will fail until we support conveyor belts. Add tests for double step conveyor belts (find out how these work at the edges)
```
Comment (kapir) [inline src/test/java/no/uib/inf112/core/map/ConveryorInteractOnUserTest.java]: Does this file test different things than conveyorTileTest or is it just outdated?

File: src/main/java/no/uib/inf112/core/map/GameMap.java [hunk 2]
```diff
@@ -62,29 +74,47 @@ public GameMap(String map) {
             throw new IllegalStateException("Given tiled map does not have a tile layer named '" + BOARD_LAYER_NAME + "'");
         }
         if (flags == null) {
-            throw new IllegalStateException("Given tiled map does not have a tile layer named '" + FLAG_LAYER_NAME + "'");
+            lasers = new TiledMapTileLayer(mapWidth, mapHeight, tileWidth, tileHeight);
```
Comment (DanielBerge) [inline src/main/java/no/uib/inf112/core/map/GameMap.java]: Why new laserlayer if one of the layers are null?

File: src/main/java/no/uib/inf112/core/map/GameMap.java [hunk 3]
```diff
@@ -110,41 +133,86 @@ public TiledMapTileSets getMapTileSets() {
...
+
+            if (laser.getX() == knownLaser.getX() && laser.getY() == knownLaser.getY()) {
+                // Already a laser tile in this layer, se if we need to change it to a cross or just ignore it (same orientation)
+                if (laser.getTile() != knownLaser.getTile()) {
+                    removeEntityLaser(knownLaser);
+                    entityLasers.add(new LaserTile(new Vector2Int(knownLaser.getX(), knownLaser.getY()), TileGraphic.LASER_CROSS, Color.WHITE));
+                    entityLaserLayer.setCell(laser.getX(), laser.getY(), new TiledMapTileLayer.Cell().setTile(TileGraphic.LASER_CROSS.getTile()));
+                    return;
+                } else {
+                    /* We should probably have a count for when this happens, cleanup will try and remove non existing tile since we didn't add it.
```
Comment (DanielBerge) [inline src/main/java/no/uib/inf112/core/map/GameMap.java]: Issue?

File: src/main/java/no/uib/inf112/core/map/MapHandler.java [hunk 4]
```diff
@@ -62,7 +66,7 @@
     void moveCamera(float dx, float dy);
 
     /**
-     * @return The cameara that renderes the board
+     * @return The camera that rendere the board
```
Comment (DanielBerge) [inline src/main/java/no/uib/inf112/core/map/MapHandler.java]: render*

File: src/main/java/no/uib/inf112/core/player/PlayerHandler.java [hunk 5]
```diff
@@ -50,76 +50,30 @@ public PlayerHandler(int playerCount, MapHandler map) {
...
-        GameGraphics.getRoboRally().getDeck().shuffle();
-        UserPlayer p = mainPlayer();
-        if (p.isPoweredDown()) {
-            //TODO Issue #24 check if is powered down (then heal)
+        Player p = (Player) mainPlayer();
+        if (p.isDestroyed()) {
             return;
+        }
+        if (p.isPoweredDown()) {
+            p.heal();
```
Comment (DanielBerge) [inline src/main/java/no/uib/inf112/core/player/PlayerHandler.java]: Heal "MAX_HEALTH" not Integer.MAXVALUE. Player should not have that kind of health

File: src/main/java/no/uib/inf112/core/map/GameMap.java (hunk 3 above)
Comment (kapir) [reply src/main/java/no/uib/inf112/core/map/GameMap.java]: Currently it is not an issue as this doesn't cause any problems, but its important for the developer to be aware of this sideeffect

File: src/test/java/no/uib/inf112/core/map/ConveryorInteractOnUserTest.java (hunk 1 above)
Comment (kh498) [reply src/test/java/no/uib/inf112/core/map/ConveryorInteractOnUserTest.java]: I have not touched this file I think,
@rikkeaas did you do anything here?

File: src/main/java/no/uib/inf112/core/map/GameMap.java (hunk 2 above)
Comment (kh498) [reply src/main/java/no/uib/inf112/core/map/GameMap.java]: This is @kapir code

File: src/main/java/no/uib/inf112/core/player/PlayerHandler.java (hunk 5 above)
Comment (kh498) [reply src/main/java/no/uib/inf112/core/player/PlayerHandler.java]: It should be capped by heal(integer), this way we can have entities with different amount of max health and they will always be healed fully

File: src/main/java/no/uib/inf112/core/map/MapHandler.java (hunk 4 above)
Comment (kapir) [reply src/main/java/no/uib/inf112/core/map/MapHandler.java]: new spelling mistake?

File: src/main/java/no/uib/inf112/core/map/GameMap.java (hunk 2 above)
Comment (kapir) [reply src/main/java/no/uib/inf112/core/map/GameMap.java]: Its not

**Pull request merged**